### PR TITLE
add checkboxes to render

### PR DESCRIPTION
**Pull request by ilyes-ced**
feature requested in issue https://github.com/markmap/markmap/issues/193

tell me what changes would be needed

and Whether the svg design is good or needs to change

and Whether to leave svg the same color as text or the color of the line

## old:
old

## new:
new

### CODE REVIEW

**Comment by gera2ld** [comment]
Thank you for your effort!

Could you make it a Markdown plugin? Here is a minimal example:

```ts
import { ITransformHooks } from '../types';
import { definePlugin } from './base';

const name = 'checkbox';

const plugin = definePlugin({
  name,
  transform(transformHooks: ITransformHooks) {
    transformHooks.parser.tap((md) => {
      md.core.ruler.before(
        'inline',
        'checkbox',
        (state) => {
          for (let i = 2; i < state.tokens.length; i += 1) {
            const token = state.tokens[i];
            if (token.type === 'inline' && token.content) {
              const prevType = state.tokens[i - 1].type;
              const prevPrevType = state.tokens[i - 2].type;
              if (prevType === 'list_item_open' || prevType === 'paragraph_open' && prevPrevType === 'list_item_open') {
                // Modify token.content
              }
            }
          }
          return false;
        },
        {},
      );
    });
    return {};
  },
});

export default plugin;
```

**Comment by gera2ld** [comment]
Also please add a test case for it, I'm afraid the current implementation matches the following cases too:

```md
[ ] this is not a checkbox

\`\`\`
[ ] this is not a checkbox either
\`\`\`
```

File: packages/markmap-lib/src/constants.ts [hunk 1]
```diff
@@ -4,3 +4,6 @@ export const baseJsPaths = [
...
 ];
+
+export const svg_marked = `<svg class="change_color" fill="currentColor" width="16px" height="16px" viewBox="0 -3 24 24" data-darkreader-inline-fill=""><g id="SVGRepo_bgCarrier" stroke-width="0"></g><g id="SVGRepo_tracerCarrier" stroke-linecap="round" stroke-linejoin="round"></g><g id="SVGRepo_iconCarrier"><path d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"></path></g></svg>`;
+export const svg_unmarked = `<svg class="change_color" fill="currentColor" width="16px" height="16px" viewBox="0 -3 24 24" xmlns="http://www.w3.org/2000/svg"><g id="SVGRepo_bgCarrier" stroke-width="0"></g><g id="SVGRepo_tracerCarrier" stroke-linecap="round" stroke-linejoin="round"></g><g id="SVGRepo_iconCarrier"> <path fill-rule="evenodd" clip-rule="evenodd" d="M6 5C5.44772 5 5 5.44772 5 6V13V18C5 18.5523 5.44772 19 6 19H18C18.5523 19 19 18.5523 19 18V13V6C19 5.44772 18.5523 5 18 5H6ZM3 6C3 4.34315 4.34315 3 6 3H18C19.6569 3 21 4.34315 21 6V13V18C21 19.6569 19.6569 21 18 21H6C4.34315 21 3 19.6569 3 18V13V6Z" data-darkreader-inline-fill=""></path> </g></svg>`;
```
Comment (gera2ld) [inline packages/markmap-lib/src/constants.ts:9]: Can you minimize the SVGs, e.g. with svgo?

**Comment by ilyes-ced** [comment]
1. svgs are minified as asked

2. checkboxes is turned to a plugin with the logic of accepting either `headers` or `list item paragraphs`
```js
// accept heading and list items paragraphs
if (prevType === 'heading_open' || (prevType === 'paragraph_open' && === 'list_item_open')) {
  if (token.content.startsWith('[ ]')) {
    token.content = `${svg_unmarked} ${token.content.slice(3)}`;
  } else if (token.content.startsWith('[x]')) {
    token.content = `${svg_marked} ${token.content.slice(3)}`;
  }
}
```

3. i tested this in other markdown formatters and those items dont get turned to checkboxes the only accepted checkboxes are `list items `

```
```md
[ ] this is not a checkbox

\`\`\`
[ ] this is not a checkbox either
\`\`\`
```

i made an exclusion to add `headings` as well because it seems usefull in this case feel free to reject it

i added test for it anyways

File: packages/markmap-lib/src/constants.ts [hunk 2]
```diff
@@ -4,3 +4,6 @@ export const baseJsPaths = [
   `d3@${process.env.D3_VERSION}/dist/d3.min.js`,
   `markmap-view@${process.env.VIEW_VERSION}/dist/browser/index.js`,
 ];
+
+export const svg_marked = `<svg width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path d="M19 3H5a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V5a2 2 0 0 0-2-2m-9 14-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8z"/></svg>`;
+export const svg_unmarked = `<svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="change_color" viewBox="0 -3 24 24"><path fill-rule="evenodd" d="M6 5a1 1 0 0 0-1 1v12a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V6a1 1 0 0 0-1-1zM3 6a3 3 0 0 1 3-3h12a3 3 0 0 1 3 3v12a3 3 0 0 1-3 3H6a3 3 0 0 1-3-3v-5z" clip-rule="evenodd"/></svg>`;
```
Comment (gera2ld) [inline packages/markmap-lib/src/constants.ts:9]: Can the `change_color` class be removed?

Otherwise it looks good to me.

**Comment by ilyes-ced** [comment]
yes i forgot that sorry

File: packages/markmap-lib/src/constants.ts (hunk 2 above)
Comment (gera2ld) [reply packages/markmap-lib/src/constants.ts]: Could you update the test case for it?

**Comment by gera2ld** [comment]
Could you fix the test case? Thanks. @ilyes-ced

**Comment by ilyes-ced** [comment]
which test case? they work fine for me and for github checks

**Comment by gera2ld** [comment]
The test case fails because the snapshot still contains `class="change_color"`. CI fails if you rebase to the latest master.
You can run `pnpm --filter markmap-lib test -- -u` to update the snapshots.

**Review by gera2ld — APPROVED** [review]
Thank you so much.

**Pull request merged**